### DISCUSSION TITLE
Is there a working sample available somewhere?

### DISCUSSION THREAD

**Discussion by richcorbs** [pinecone-router/router#5] [Q&A] [answered]
This looks interesting but I'm struggling to understand how the pieces go together.

Is there a working example somewhere?

**Comment by rehhouari** [comment]
Hey, not yet sorry! I'll set up one very soon and link it!
For now here's how to do basic route handling
import the alpine router *before* alpine, then just use the example in the readme as well:
(Assuming you know how to make the webserver you're using to route things to index.html, if not or you just cant, use hash routing by adding `x-hash` to the router element)
```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<script src="https://cdn.jsdelivr.net/gh/rehhouari/alpine-router@0.0.4/dist/complete.umd.js"></script>
		<script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.x.x/dist/alpine.min.js" defer></script>
	</head>

	<body>
		<script>
			function handle() {
				return {
					main() {
						console.log('main');
					},
					hello(context) {
						console.log('hello,', context.props.name);
					},
					notfound(context) {
						console.error(context.path + ' is not found');
					}
				};
			}
		</script>
		<div x-data="handle()" x-router>
			<template x-route="/hello/:name" x-handler="hello"></template>
			<template x-route="/" x-handler="main"></template>
			<template x-route="notfound" x-handler="notfound"></template>
		</div>
		<nav class="">
			<a href="/">Home</a>
			<a href="/hello/test">Hello</a>
			<a href="/about">About</a>
		</nav>
	</body>
</html>
```

**Reply by richcorbs** [reply]
Thanks for this.

I assume I need to also create `/hello.html`. Is that correct?

The links in the nav work to change the URL but how/where do I put the content for the "hello" page?

Do I need to something in the handler function to display the new page content?

**Reply by rehhouari** [reply]
Oh you mean that! I'll explain it since it isn't clear at all for someone who haven't use client side routing before :)

Alpine Router give you 3 options:

### First option, using x-handler

You don't need to have a file `/hello.html` but yes you need to have something in the handler to display page or do anything you want, it dont have to be displaying. for example the hanlder here in the example prints things to the *console*

**How?** clicking the link don't visit `/hello` on the server, it simply _change the url_ and call the _handler_. so this give you access to the current _route_ and you do _whatever you want_ with it!

**Why?** you may have your own method of displaying the content, or you dont want that you want to do something else you can and you get access to the [context](https://github.com/rehhouari/alpine-router/#global-context) in the handler to help you.

### Second option, x-views

You want to display content but not an entire page, just a part, for example you have "/profile.html" you want to show it if  the user want to view someones profile you can use this:
```html
<div x-data x-router x-views>
  <template x-route="/profile/:name" x-view="profile.html"></template>
 <template x-route="notfounf" x-view="notfound.html"></template>
</div>
<div id="content">
This here will be replaced by the content of `profile.html` if the user visit `/profile/john`.
If the user visit _any path_ but that, it'll show the content of `notfound.html`
</div>
```
notice that there is no need for `x-handler` this way, but you can add it and it'll be executed.
by default this will display the content in the element that have `id=content`, it can be ovewritten by setting `x-views="#some-other-id"`
**Note:** you can have your routes share view files, which are normal html files. they can also contain Alpine.js components and they will work normally

### Third option, x-render

This will let you render pages in the server, normally but with a twist, You don't need to handle routes or anything, just add this line:
```html
<div x-data x-router x-views></div>
```
What this will do:

- whenever you hover over a relative link (within the site) for a little less than a second, it  preloads its content (full HTML Page).
- when you click on the link it display that _preloaded_ content replacing the entire `<body>` instead of the entire document.
-
This result in faster page navigation with no reloads! the server generate the html, the first load is normal, meaning it's SEO friendly. This method is made popular by [Turbolinks](https://github.com/turbolinks/turbolinks)

**Comment by rehhouari** [comment, marked as answer]
there is now: https://pinecone-example.vercel.app
